Update default transport to use clone feature

kconwayinvision opened the pull request:
This helps ensure the OptionDefaultTransport is always up to date by
leveraging the built-in Clone() method of the *http.Transport.

The added test is a basic check to ensure that all the comparable values
match. Additional fields may appear over time but this _should_ be
enough to prove that the Clone() feature is working as expected.

kconwayinvision commented:
Note that this is slightly different behavior than before because the transport is being overwritten entirely with the default values rather than modified to match. There's not a way to keep the copied fields up to date without doing this. An alternative may be to deprecate the `OptionDefaultTransport` and, instead, use the `Clone()` method to generate the initial transport rather than use an empty transport.

Also note that `Clone` is a feature added after 1.12.4 which is currently the pinned Go version for the SDCLI build tool.

kconwayinvision commented:
https://github.com/asecurityteam/sdcli/pull/31 for a PR that updates the SDCLI build tool to work with this change.

kconwayinvision commented:
> Info: Even though the implementation is now almost a no-op(), I think leaving this as an `Option` type provides a nicer interface to consumers, so I prefer this approach vs deprecating `OptionDefaultTransport` in favor of callers invoking `Clone()` themselves.

To clarify, I meant moving the `Clone()` call to within the `New()` method rather than having it start with a `&http.Transport{}`. For example:

```golang
unc New(opts ...Option) *http.Transport {
	var t = http.DefaultTransport.(*http.Transport).Clone()
	for _, opt := range opts {
		t = opt(t)
	}
	return t
}
```

This would still keep the `.Clone()` call hidden from the user while also potentially making an empty `New()` call more valuable by not defaulting to all zero values (infinite timeouts, etc.). It would still be a change in behavior, though.

One general problem is that both the current PR and the alternative introduce a potentially breaking change in the lib. For example, the previous version applied additional options onto a potentially partially configured transport. The new version returns the default transport in a way that would not account for any existing configuration. For example, the following might no longer work as expected after the PR:

```golang
t := transport.New(OptionDisableKeepalives(true), OptionDefaultTransport)
```

If your expectation was that the `OptionDefaultTransport` applies all non-zero value configuration on top of the `OptionDisableKeepalives(true)` then it is not broken because the new version will apply _all_ default configuration values. If, however, you think the method was always intended to apply _all_ default configuration values then the PR fixes a bug where it did not do that.

I think the only way to do this in a truly non-breaking change would be to introduce a new option for the new behavior and/or provide a new `New` that uses `Clone()` internally like a `NewFromDefault(...)`.

Do any of those ideas sound appealing?

faldridge commented:
Ah, ok, thanks for the clarification!

This is a tough one, but I do actually prefer the behavior of `New` never returning a potentially partially configured transport. 

Suggestion: call this one a bug fix and implement it with `Clone` in the `New` function, deprecating the use of `OptionDefaultTransport`.

kconwayinvision commented:
@faldridge I've moved the clone into the `New` method and marked the old option as deprecated which should both remove it from godocs and emit a linter warning to users. I added some tests to cover the new default behavior in both the old option and the `New` method.

faldridge commented:
Thanks so much, @kconwayinvision! One more hopefully minor request, and I do apologize for the back-and-forth:

Task: Please update the README to stop using the deprecated 
`transport.OptionDefaultTransport` `Option` in the usage example?

kconwayinvision commented:
@faldridge  Good call-out on the README. I've removed all references to the deprecated option. The nice thing is that all of the example code still has the same behavior without the option being added so the overall text is still accurate.
